### PR TITLE
*: add writer to backup manager 

### PR DESCRIPTION
**Pull request by fanminshi**
* backup manager now can write backup to a specific path prefix with the `SaveSnapWithPrefix` function.
* add NewBackupManagerFromWriter factory function to create a BackupManager based on backup writer.
* backup operator uses `SaveSnapWithPrefix` to save snapshot.

### CODE REVIEW

File: pkg/backup/backup_manager.go [hunk 1]
```diff
@@ -56,12 +58,64 @@ func NewBackupManager(kubecli kubernetes.Interface, clusterName string, namespac
 	}
 }
 
+// NewBackupManagerFromWriter creates a BackupManager with backup writer.
+func NewBackupManagerFromWriter(kubecli kubernetes.Interface, clusterName string, namespace string, etcdTLSConfig *tls.Config, bw writer.Writer) *BackupManager {
```
Comment (hongchaodeng) [inline pkg/backup/backup_manager.go]: `clusterName, namespace string`

**Comment by fanminshi** [comment]
cc/ @hongchaodeng @hasbro17

**Comment by fanminshi** [comment]
all fixed. PTAL @hongchaodeng @hasbro17

File: pkg/backup/backup_manager.go [hunk 2]
```diff
@@ -79,6 +92,25 @@ func (bm *BackupManager) SaveSnap(lastSnapRev int64) (*backupapi.BackupStatus, e
 	return bs, nil
 }
 
+// SaveSnapWithPath saves latest snapshot to a given path.
+func (bm *BackupManager) SaveSnapWithPath(path string) error {
```
Comment (hongchaodeng) [inline pkg/backup/backup_manager.go]: Where is this used?

Comment (fanminshi) [reply pkg/backup/backup_manager.go]: I am think that this should be used in backup operator to save snapshot to a specific path.

Comment (hongchaodeng) [reply pkg/backup/backup_manager.go]: Can you submit the changes for backup operator using it?

Comment (fanminshi) [reply pkg/backup/backup_manager.go]: do you want me to include in this pr?

Comment (hongchaodeng) [reply pkg/backup/backup_manager.go]: Yes, please

Comment (fanminshi) [reply pkg/backup/backup_manager.go]: @hongchaodeng should we also change backup spec to take in a path too?

Comment (hongchaodeng) [reply pkg/backup/backup_manager.go]: Not right now.

**Comment by fanminshi** [comment]
all fixed. PTAL cc/ @hongchaodeng @hasbro17

File: pkg/backup/backup_manager.go [hunk 3]
```diff
@@ -113,6 +127,49 @@ func (bm *BackupManager) writeSnap(mcli clientv3.Maintenance, endpoint string, r
 	return bs, nil
 }
 
+// SaveSnapWithPrefix saves latest snapshot to a path prepended with the given prefix
+// and returns file size and full path.
+// the full path has the format of prefix/<etcd_version>_<snapshot_reversion>_etcd.backup
+// e.g prefix = etcd-backups/v1/default/example-etcd-cluster and
+// backup object name = 3.1.8_0000000000000001_etcd.backup
+// "etcd-backups/v1/default/example-etcd-cluster/3.1.8_0000000000000001_etcd.backup".
```
Comment (hongchaodeng) [inline pkg/backup/backup_manager.go]: full path is "..."

File: pkg/backup/backup_manager.go [hunk 4]
```diff
@@ -113,6 +127,49 @@ func (bm *BackupManager) writeSnap(mcli clientv3.Maintenance, endpoint string, r
...
+	}
+	fullPath := path.Join(prefix, util.MakeBackupName(version, rev))
+	n, err := bm.bw.Write(fullPath, rc)
+	if err != nil {
+		return 0, "", fmt.Errorf("failed to write snapshot (%v)", err)
+	}
+	return n, fullPath, nil
+}
+
+func etcdVersion(etcdcli *clientv3.Client) (string, error) {
```
Comment (hongchaodeng) [inline pkg/backup/backup_manager.go]: getEtcdVersion

File: pkg/backup/backup_manager.go [hunk 5]
```diff
@@ -113,6 +127,49 @@ func (bm *BackupManager) writeSnap(mcli clientv3.Maintenance, endpoint string, r
...
+	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
+	resp, err := etcdcli.Status(ctx, etcdcli.Endpoints()[0])
```
Comment (hongchaodeng) [inline pkg/backup/backup_manager.go]: Can you reuse it in writeSnap() too?

File: pkg/backup/backup_manager.go [hunk 6]
```diff
@@ -113,6 +127,49 @@ func (bm *BackupManager) writeSnap(mcli clientv3.Maintenance, endpoint string, r
...
+	cancel()
+	if err != nil {
+		return "", fmt.Errorf("failed to receive etcd reversion (%v)", err)
```
Comment (hongchaodeng) [inline pkg/backup/backup_manager.go]: reversion  -> version

File: pkg/controller/backup-operator/s3_backup.go [hunk 7]
```diff
@@ -37,23 +35,12 @@ func handleS3(kubecli kubernetes.Interface, s3 *api.S3Source, namespace, cluster
...
-	prefix := backupapi.ToS3Prefix(s3.Prefix, namespace, clusterName)
-	s3cli := backups3.NewFromClient(s3.S3Bucket, prefix, cli.S3)
-	be := backend.NewS3Backend(s3cli)
-	bm := backup.NewBackupManager(kubecli, clusterName, namespace, nil, be)
-
-	// this SaveSnap takes 0 as lastSnapRev to indicate that it
-	// saves any snapshot with revision > 0.
-	bs, err := bm.SaveSnap(0)
+	// TODO: support TLS.
+	bm := backup.NewBackupManagerFromWriter(kubecli, nil, writer.NewS3Writer(cli.S3), clusterName, namespace)
```
Comment (hongchaodeng) [inline pkg/controller/backup-operator/s3_backup.go]: Let's remove the etcdTLSConfig field for now? It already has a TODO

File: pkg/controller/backup-operator/s3_backup.go [hunk 8]
```diff
@@ -37,23 +35,12 @@ func handleS3(kubecli kubernetes.Interface, s3 *api.S3Source, namespace, cluster
...
+	s3Prefix := backupapi.ToS3Prefix(s3.Prefix, namespace, clusterName)
+	_, fullPath, err := bm.SaveSnapWithPrefix(path.Join(s3.S3Bucket, s3Prefix))
```
Comment (hongchaodeng) [inline pkg/controller/backup-operator/s3_backup.go]: No need to return Size?

Comment (fanminshi) [reply pkg/controller/backup-operator/s3_backup.go]: size is not used anywhere. so no need to return it.

**Comment by fanminshi** [comment]
all fixed. PTAL cc/ @hongchaodeng @hasbro17

**Comment by fanminshi** [comment]
all fixed. PTAL cc/ @hongchaodeng @hasbro17

**Comment by fanminshi** [comment]
all fixed. PTAL cc/ @hongchaodeng @hasbro17

**Pull request merged**